File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Medium/36. Valid Sudoku/C++/36. Valid Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<unordered_set<char>> rows(9), cols(9);
        vector<unordered_set<char>> subgrids(9);

        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.') continue; 

                char num = board[i][j];
                int subgridIndex = (i / 3) * 3 + (j / 3);

                if (rows[i].count(num) || cols[j].count(num) || subgrids[subgridIndex].count(num)) {
                    return false;
                }

                rows[i].insert(num);
                cols[j].insert(num);
                subgrids[subgridIndex].insert(num);
            }
        }
        return true;
    }
};

File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Medium/36. Valid Sudoku/Python/36. Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = defaultdict(set)
        col = defaultdict(set)
        s = defaultdict(set)

        for i in range(9):
            for j in range(9):
                if board[i][j] == ".": continue

                num = board[i][j]
                if num in row[i] or num in col[j] or num in s[i//3, j//3]:
                    return False
                
                row[i].add(num)
                col[j].add(num)
                s[i//3, j//3].add(num)
                
        return True


                